### DISCUSSION TITLE
Aptos Indexer Support

### DISCUSSION THREAD

**Discussion by l-monninger** [movementlabsxyz/movement#249] [Ideas]
# Summary
Guidelines for supporting the Aptos Indexer framework.

## Background
- [Aptos Indexer](https://aptos.dev/en/build/indexer) architecture and framework
   - The [Transaction Streaming Service](https://aptos.dev/en/build/indexer/txn-stream) is the most important component for people looking to build `indexer-processors` atop the current indexer. We have enabled the TSS via a service served directly on the full-node, though not in the manner the full implementation suggests. See item bullet 2 on #210 . 
   - The [GraphQL API](https://aptos.dev/en/build/indexer/api/self-hosted) is the most important for dApp developers who don't need their own custom business logic in an `indexer-processor`. 
   - At various points in time Aptos has provided notes that their GraphQL service implementation is closed source. They provide a Docker image that you can try to use. 
   
## Next Steps
- The easiest place to start is simply by seeing whether the docker image works with our existing TSS. If it doesn't go forth and implement the more complete TSS, i.e., setting the `node_config.indexer_grpc.use_data_service_interface = true;` to false and starting a separate service.
- If this doesn't work, you'll need to write your own GraphQL wrapper for the DB. You'll want to write this in Rust, so that you can use the existing types.

**Comment by l-monninger** [comment]
Since a couple days ago, the `postgres` service in the `docker-compose` application is not made correctly available to the full-node. To work around this, you may docker run the Aptos `graphql` container within process compose.

**Comment by andyjsbell** [comment]
I had a peek at this and it seems we can take https://github.com/aptos-labs/aptos-indexer-processors/ and point it to our `data_service` and the `postgres`.  There are processors in `rust` and `python`

**Reply by andyjsbell** [reply]
the `data_service` would be on `localhost:30734`

**Reply by l-monninger** [reply]
Yes, the processors should be compatible since we synced with the upstream. If you need to make changes, make them here: https://github.com/movementlabsxyz/aptos-indexer-processors/tree/l-monninger/indexer-instrumentation

However, this will not give your GraphQL API. For that, Aptos is using Hasura to not have to build anything out themselves. We can do the same with the schemas available instead of writing anything in `juniper` or similar. 

My comment above about them providing a Docker image for running a GraphQL API over Postgres is misleading, I think. Instead I believe they're just using Hasura's docker setup: https://hasura.io/docs/latest/getting-started/docker-simple/
